### PR TITLE
Ticket31684

### PR DESCRIPTION
**Pull request by ltbringer**
Add control port GETINFO support for dumping the local consensus

Refer: (ticket)[https://trac.torproject.org/projects/tor/ticket/31684]

### CODE REVIEW

File: changes/ticket31684 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+  o Minor features (onion services):
+    - GETINFO dir/status-vote/current/consensus-microdesc dumps microdesc consensus. Closes ticket 31684.
```
Comment (asn-d6) [inline changes/ticket31684]: Please check how other changes entries look like. They are usually human readable and look like "Implement a new GETINFO command ...".

File: src/feature/control/control_getinfo.c [hunk 2]
```diff
@@ -325,6 +325,51 @@ getinfo_helper_current_time(control_connection_t *control_conn,
   return 0;
 }
 
+/**
+ * Switch between microdesc vs networkstatus descriptor dumps
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Please improve the function documentation here by making it more readable. Please see how the other GETINFO helper functions are documented and take a similar human readable approach.

Also please document the return value of the function.

File: src/feature/control/control_getinfo.c [hunk 3]
```diff
@@ -325,6 +325,51 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+ * - dir/status-vote/current/consensus-microdesc
+ * or
+ * - dir/status-vote/current/consensus
+ */
+STATIC int
+getinfo_helper_current_consensus(const char* question,
+                          char** answer,
+                          const char** errmsg)
+{
+  // Ensures question contains a request for
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Please use the `/*` `*/` comment style for comments.

File: src/feature/control/control_getinfo.c [hunk 4]
```diff
@@ -325,6 +325,51 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+      strcmp(question, "dir/status-vote/current/consensus") &&
+      strcmp(question, "dir/status-vote/current/consensus-microdesc")) {
+    return 0;
+  }
+
+  const char* consensus_type = !strcmp(
+     question, "dir/status-vote/current/consensus-microdesc"
+  ) ? "microdesc" : "ns";
+
+  if (we_want_to_fetch_flavor(get_options(), FLAV_NS)) {
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Hmm, is this right? We call the `we_want_to_fetch_flavor()` with `FLAV_NS` but then we call `dirserv_get_consensus()` with arbitrary `consensus_type`?

Also, can you please document what's the purpose of this `we_want_to_fetch_flavor()` code block here?

File: src/feature/control/control_getinfo.c [hunk 5]
```diff
@@ -362,6 +407,7 @@ getinfo_helper_dir(control_connection_t *control_conn,
       if (body)
         *answer = tor_strndup(body, ri->cache_info.signed_descriptor_len);
     } else if (! we_fetch_router_descriptors(get_options())) {
+
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Newline added for no real reason.

File: src/feature/control/control_getinfo.c [hunk 6]
```diff
@@ -576,24 +622,16 @@ getinfo_helper_dir(control_connection_t *control_conn,
...
-        *answer = tor_memdup_nulterm(mapped->data, mapped->size);
-        tor_munmap_file(mapped);
-      }
-      if (!*answer) { /* generate an error */
-        *errmsg = "Could not open cached consensus. "
-          "Make sure FetchUselessDescriptors is set to 1.";
-        return -1;
-      }
-    }
+  } else if (-1 == getinfo_helper_current_consensus(
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: I'm not sure if I like the interface here:
- If you see all the other checks in `getinfo_helper_dir()` you will notice that they only get in the `if` block if the GETINFO command matches, whereas in this patch we get in the block if there is an error. I think this is not the way it should be done, and we should copy the approach of the rest of the function. Also:

- I'd prefer if we explicitly checked the GETINFO string in this function like above, instead of hiding it into the helper function. I would probably go with:
  ```
  ...
  } else if (!strcmp(question, "dir/status-vote/current/consensus")) { /* ns consensus */
      get_info_helper_current_consensus("ns");
  } else if (!strcmp(question, "dir/status-vote/current/consensus-microdesc)) { /* md consensus */
      get_info_helper_current_consensus("nicrodesc");
  ...
  ```

- We don't usually use Yoda conditions in Tor.

File: src/feature/control/control_getinfo.c [hunk 7]
```diff
@@ -576,24 +622,16 @@ getinfo_helper_dir(control_connection_t *control_conn,
...
+                                                    question,
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Why intend the arguments in 4 different lines? If they fit, just put them in the same line except if you have something in mind.

File: src/feature/control/control_getinfo.c [hunk 8]
```diff
@@ -325,6 +325,51 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+ * - dir/status-vote/current/consensus
+ */
+STATIC int
+getinfo_helper_current_consensus(const char* question,
+                          char** answer,
+                          const char** errmsg)
+{
+  // Ensures question contains a request for
+  // ns or microdesc consensus
+  if (
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Why is there a gap here? You could put the `strcmp` in the `if` line, except if you have something specific in mind.

File: src/feature/control/control_getinfo.c [hunk 9]
```diff
@@ -1513,6 +1551,8 @@ static const getinfo_item_t getinfo_items[] = {
          "v2 networkstatus docs as retrieved from a DirPort."),
   ITEM("dir/status-vote/current/consensus", dir,
        "v3 Networkstatus consensus as retrieved from a DirPort."),
+  ITEM("dir/status-vote/current/consensus-microdesc", dir,
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Please consider adding a unittest for the new functionality. See how `test_download_status_consensus()` does it for example.

File: src/feature/control/control_getinfo.c (hunk 7 above)
Comment (ltbringer) [reply src/feature/control/control_getinfo.c]: There is a lint error that restricts lines to some 78 chars. That's why I thought of breaking them into different lines.

File: src/feature/control/control_getinfo.c (hunk 6 above)
Comment (ltbringer) [reply src/feature/control/control_getinfo.c]: I understand, I got stuck thinking about how to move things out from the conditionals. I'll fix it.

File: src/feature/control/control_getinfo.c [hunk 10]
```diff
@@ -326,46 +326,33 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
-
-  const char* consensus_type = !strcmp(
-     question, "dir/status-vote/current/consensus-microdesc"
-  ) ? "microdesc" : "ns";
 
-  if (we_want_to_fetch_flavor(get_options(), FLAV_NS)) {
-    const cached_dir_t *consensus = dirserv_get_consensus(consensus_type);
-    if (consensus) {
-      *answer = tor_strdup(consensus->dir);
-    }
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Are we sure we don't need the `we_want_to_fetch_flavor()` call here anymore? What purpose was it serving in the past?

File: src/test/test_controller.c [hunk 11]
```diff
@@ -1689,6 +1689,42 @@ test_download_status_bridge(void *arg)
...
+  control_connection_t dummy;
+  /* Get results out */
+  char *answer = NULL;
+  const char *errmsg = NULL;
+
+  (void)arg;
+
+  setup_bridge_mocks();
+
+  getinfo_helper_dir(&dummy,
```
Comment (asn-d6) [inline src/test/test_controller.c]: Nice start! However this unittest does not actually test that the right consensus is returned. Ideally the unittest here should make sure that the right consensus string will be returned. (Hint: This might require some mocking)

File: src/feature/control/control_getinfo.c (hunk 10 above)
Comment (ltbringer) [reply src/feature/control/control_getinfo.c]: I checked `we_want_to_fetch_flavor()` function and I saw that it returns a list of consensus flavors. since we are passing that ourselves, I thought about removing it.

Comment (teor2345) [reply src/feature/control/control_getinfo.c]: I think the we_want_to_fetch_flavor() check is used to make sure that the flavour is in RAM or on disk.

File: src/feature/control/control_getinfo.c (hunk 4 above)
Comment (teor2345) [reply src/feature/control/control_getinfo.c]: I think the we_want_to_fetch_flavor() check is used to make sure that the flavour is in RAM or on disk.

File: src/feature/control/control_getinfo.c [hunk 12]
```diff
@@ -326,46 +326,37 @@ getinfo_helper_current_time(control_connection_t *control_conn,
 }
 
 /**
- * Switch between microdesc vs networkstatus descriptor dumps
- * Assumes that question would be either:
- * - dir/status-vote/current/consensus-microdesc
- * or
- * - dir/status-vote/current/consensus
+ * Helps to access directory information for a given
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: This still doesn't look like our usual documentation format. Please describe the return value in a proper sentence and not in a list.

File: src/feature/control/control_getinfo.c [hunk 13]
```diff
@@ -326,46 +326,37 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
       *answer = tor_strdup(consensus->dir);
     }
   }
-  if (!*answer) { /* try loading it from disk */
+  if (!*answer) { /* else, try loading consensus from disk */
     tor_mmap_t *mapped = networkstatus_map_cached_consensus(consensus_type);
     if (mapped) {
       *answer = tor_memdup_nulterm(mapped->data, mapped->size);
       tor_munmap_file(mapped);
+      tor_free(mapped);
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Why are you freeing this here? Are you sure this is right? Please double check this because I think you might be freeing the actual global consensus.

File: src/feature/control/control_getinfo.c [hunk 14]
```diff
@@ -622,16 +613,17 @@ getinfo_helper_dir(control_connection_t *control_conn,
...
-                                                    answer,
-                                                    errmsg))
-  {
-    /**
-     * answer is set by getinfo_helper_current_consensus
-     * if the question matches
-     */
-    return -1;
+  } else if (!strcmp(question, "dir/status-vote/current/consensus")) {
+    if (getinfo_helper_current_consensus(
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Please don't wrap the code like this. Check the rest of the code, we usually wrap it between the arguments and not after the parentheses.

We also usually do `< 0` when checking for errors and not `== -1`.

File: src/test/test_controller.c [hunk 15]
```diff
@@ -1689,6 +1695,62 @@ test_download_status_bridge(void *arg)
...
+}
+
+static void
+test_getinfo_helper_dir(void *arg)
+{
+  /* We just need one of these to pass, it doesn't matter what's in it */
+  control_connection_t dummy;
+  /* Get results out */
+  char *answer = NULL;
+  const char *errmsg = NULL;
```
Comment (asn-d6) [inline src/test/test_controller.c]: This unittest is failing in Appveyor for Windows. Any idea why? It might be that the extra free is causing trouble, or something else.

File: src/feature/control/control_getinfo.c [hunk 16]
```diff
@@ -326,46 +326,37 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
-  }
-
-  const char* consensus_type = !strcmp(
-     question, "dir/status-vote/current/consensus-microdesc"
-  ) ? "microdesc" : "ns";
-
-  if (we_want_to_fetch_flavor(get_options(), FLAV_NS)) {
+  /** Check if the consensus flavor is cached or in the ram. */
+  if (we_want_to_fetch_flavor(get_options(), flavor)) {
     const cached_dir_t *consensus = dirserv_get_consensus(consensus_type);
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Can you also test this branch in the unittest perhaps?

Comment (ltbringer) [reply src/feature/control/control_getinfo.c]: Yes I can try.

File: src/feature/control/control_getinfo.c (hunk 13 above)
Comment (ltbringer) [reply src/feature/control/control_getinfo.c]: The CI was complaining about memory leaks and that's why I tried to do this. I found that the problem was in the test itself where I wasn't freeing `answer` I'll undo this and check.

File: src/test/test_controller.c (hunk 15 above)
Comment (ltbringer) [reply src/test/test_controller.c]: It is something else :( I haven't been able to find what exactly. Without the extra free here the CI was failing on both Travis and Appveyor due to memory leaks.

File: src/feature/control/control_getinfo.c [hunk 17]
```diff
@@ -325,6 +325,33 @@ getinfo_helper_current_time(control_connection_t *control_conn,
   return 0;
 }
 
+/**
+ * Helps to access directory information for a given
+ * type of consensus.
+ *
+ * returns: int
```
Comment (teor2345) [inline src/feature/control/control_getinfo.c]: `returns: int` is redundant, because the code tells us the type that the function returns.

File: changes/ticket31684 [hunk 18]
```diff
@@ -0,0 +1,2 @@
+  o Minor features (onion services):
+    - Implement a new GETINFO command to fetch consensus. Closes ticket 31684.
```
Comment (teor2345) [inline changes/ticket31684]:
```suggestion
    - Implement a new GETINFO command to fetch the microdesc consensus. Closes ticket 31684.
```

File: src/feature/control/control_getinfo.h [hunk 19]
```diff
@@ -48,6 +48,10 @@ STATIC int getinfo_helper_downloads(
     control_connection_t *control_conn,
     const char *question, char **answer,
     const char **errmsg);
+STATIC int getinfo_helper_current_consensus(
+    const char* consensus_type,
```
Comment (teor2345) [inline src/feature/control/control_getinfo.h]: Other code uses flavor_name
```suggestion
    const char* flavor_name,
```

File: src/feature/control/control_getinfo.c [hunk 20]
```diff
@@ -325,6 +325,33 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+ *   -1 in cases of an error
+ *   0 otherwise
+ */
+STATIC int
+getinfo_helper_current_consensus(const char* consensus_type,
```
Comment (teor2345) [inline src/feature/control/control_getinfo.c]: Other code uses flavor_name, please be consistent

File: src/feature/control/control_getinfo.c [hunk 21]
```diff
@@ -325,6 +325,33 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+                                 char** answer,
+                                 const char** errmsg)
+{
+  tor_mmap_t *mapped = networkstatus_map_cached_consensus(consensus_type);
+  if (mapped) {
+    *answer = tor_memdup_nulterm(mapped->data, mapped->size);
+    tor_munmap_file(mapped);
+    tor_free(mapped);
```
Comment (teor2345) [inline src/feature/control/control_getinfo.c]: Double-free, this line is not in the original code at:
https://github.com/torproject/tor/pull/1328/files#diff-ac2399586b806edb39db242c0a5664bbL589

File: src/feature/control/control_getinfo.c [hunk 22]
```diff
@@ -325,6 +325,33 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+ *
+ * returns: int
+ *   -1 in cases of an error
+ *   0 otherwise
+ */
+STATIC int
+getinfo_helper_current_consensus(const char* consensus_type,
+                                 char** answer,
+                                 const char** errmsg)
+{
```
Comment (teor2345) [inline src/feature/control/control_getinfo.c]: Why did you delete the code that checks for a consensus in memory in this refactor?
```
    if (we_want_to_fetch_flavor(get_options(), FLAV_NS)) {
      const cached_dir_t *consensus = dirserv_get_consensus("ns");
      if (consensus)
        *answer = tor_strdup(consensus->dir);
    }
    if (!*answer) { /* try loading it from disk */
```

Comment (ltbringer) [reply src/feature/control/control_getinfo.c]: I am checking for memory leaks in the code, I am sorry these were not meant for review. I shall have that code back once I understand what is causing the seg fault and memory leak on mingw and clang.

**Comment by teor2345** [comment]
It seems like you're pushing a lot of commits here.
Can we help you get set up to run these tests locally?

Maybe this task isn't a good fir for you.
Would you like us to take over this refactor?

**Comment by ltbringer** [comment]
I'll take it off on a separate branch and use the CI on my fork. I assumed I'd get this done faster but I should have done this before. Sincere apologies.

File: changes/ticket31684 [hunk 23]
```diff
@@ -1,2 +1,5 @@
   o Minor features (onion services):
-    - GETINFO dir/status-vote/current/consensus-microdesc dumps microdesc consensus. Closes ticket 31684.
+    - Implement a new GETINFO command to fetch microdesc consensus. Closes ticket 31684.
```
Comment (asn-d6) [inline changes/ticket31684]: "fetch the microdescriptor consensus"

File: src/feature/control/control_getinfo.c [hunk 24]
```diff
@@ -325,47 +325,32 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
 }
 
-/**
- * Switch between microdesc vs networkstatus descriptor dumps
- * Assumes that question would be either:
- * - dir/status-vote/current/consensus-microdesc
- * or
- * - dir/status-vote/current/consensus
- */
+/** Fetch network-status or micro-descriptor
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: "GETINFO helper for dumping different consensus flavors".

File: src/feature/control/control_getinfo.c [hunk 25]
```diff
@@ -325,47 +325,32 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+ * returns: 0 on success -1 on error. */
 STATIC int
-getinfo_helper_current_consensus(const char* question,
-                          char** answer,
-                          const char** errmsg)
+getinfo_helper_current_consensus(int flavor,
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: No need to pass both `flavor` and `flavor_name`. That's redundant.

Please pass just one, and derive the other in the function.

File: src/test/test_controller.c [hunk 26]
```diff
@@ -1689,6 +1691,187 @@ test_download_status_bridge(void *arg)
   return;
 }
 
+/** Mock cached consensus */
+static cached_dir_t *mock_ns_consensus_cache;
+static cached_dir_t *mock_microdesc_consensus_cache;
+
+int
```
Comment (asn-d6) [inline src/test/test_controller.c]: Why do you have the function prototype and the actual function separate? We don't do that. Please dump the function prototype and just document the function directly. Same for the ones below.

File: src/test/test_controller.c [hunk 27]
```diff
@@ -1689,6 +1691,187 @@ test_download_status_bridge(void *arg)
...
+mock_networkstatus_get_cache_fname(int flav,
+                                   const char *flavor_name,
+                                   int unverified_consensus);
+
+char *
+mock_networkstatus_get_cache_fname(int flav,
+                                   const char *flavor_name,
+                                   int unverified_consensus)
+{
+  /** A function needs to do something with
```
Comment (asn-d6) [inline src/test/test_controller.c]: You don't need to do useless stuff with variables. Instead just do `(void) flav` if you have no need for the argument.

File: src/test/test_controller.c [hunk 28]
```diff
@@ -1689,6 +1691,187 @@ test_download_status_bridge(void *arg)
...
+mock_we_want_to_fetch_flavor(const or_options_t *options,
+                             int flavor) {
+  if (options->FetchUselessDescriptors ||
+      flavor == FLAV_NS || flavor == FLAV_MICRODESC) {
+    return 1;
+  }
+  return 0;
+}
+
+/**  Mock the function that retrieves consensus from
```
Comment (asn-d6) [inline src/test/test_controller.c]: Seems like you are wrapping text at like 60 columns, whereas we usually wrap at 80. It seems even more weird than our already weird wrapping logic, so please adjust all your code to fit the norm.

File: src/test/test_controller.c [hunk 29]
```diff
@@ -1689,6 +1691,187 @@ test_download_status_bridge(void *arg)
...
+  /* Get results out */
+  char *answer = NULL;
+  const char *errmsg = NULL;
+
+  (void)arg;
+
+  setup_bridge_mocks();
+  MOCK(tor_mmap_file, mock_tor_mmap_file);
+  MOCK(tor_munmap_file, mock_tor_munmap_file);
+  MOCK(networkstatus_get_cache_fname,
```
Comment (asn-d6) [inline src/test/test_controller.c]: I don't think you need to mock `networkstatus_get_cache_fname()`. Just use the `get_options_mutable()` to change the options so that the function does not crash. See how other tests are doing it. e.g. `test_router_dump_router_to_string_no_bridge_distribution_method()`.

File: src/test/test_controller.c [hunk 30]
```diff
@@ -1689,6 +1691,187 @@ test_download_status_bridge(void *arg)
...
+  /* We just need one of these to pass, it doesn't matter what's in it */
+  control_connection_t dummy;
+  /* Get results out */
+  char *answer = NULL;
+  const char *errmsg = NULL;
+
+  (void)arg;
+
+  setup_bridge_mocks();
+  MOCK(we_want_to_fetch_flavor, mock_we_want_to_fetch_flavor);
```
Comment (asn-d6) [inline src/test/test_controller.c]: Why do you mock `we_want_to_fetch_flavor()`? This seems like an interesting function that we should know how to navigate, instead of black-boxing it. Please try to navigate it by using `or_options_t` as suggested above.

File: src/test/test_controller.c [hunk 31]
```diff
@@ -1838,6 +2020,10 @@ struct testcase_t controller_tests[] = {
     NULL },
   { "download_status_consensus", test_download_status_consensus, 0, NULL,
     NULL },
+  {"getinfo_helper_dir_from_cache", test_getinfo_helper_dir_from_cache,
```
Comment (asn-d6) [inline src/test/test_controller.c]: Please rename these tests to clarify that they test the consensus fetching function and not the whole `getinfo_helper_dir()`.

File: src/test/test_controller.c (hunk 26 above)
Comment (ltbringer) [reply src/test/test_controller.c]: These are not functions, these are variables that I create outside the test case so that I can free them. The original code:

```
  if (we_want_to_fetch_flavor(get_options(), flavor)) {
    /** Check from the cache */
    const cached_dir_t *consensus = dirserv_get_consensus(flavor_name);
    if (consensus) {
      *answer = tor_strdup(consensus->dir);
    }
  }
```
... doesn't free the consensus as it points to a global object. I keep these mock variables for the sake of avoiding memory leaks from within the test by freeing them after use.

File: src/test/test_controller.c (hunk 27 above)
Comment (ltbringer) [reply src/test/test_controller.c]: Thanks, I didn't know I could do that.

File: src/test/test_controller.c (hunk 29 above)
Comment (ltbringer) [reply src/test/test_controller.c]: I am trying to mock the parts of [this function](https://github.com/torproject/tor/blob/35cfe2e77679e1df5c3ae4b71112fa252acafdca/src/feature/nodelist/networkstatus.c#L245)

I don't see it using an `or_options_t` nor its callee `networkstatus_map_cached_consensus` ?

File: src/feature/control/control_getinfo.c [hunk 32]
```diff
@@ -325,47 +325,37 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
- * or
- * - dir/status-vote/current/consensus
- */
+/** GETINFO helper for dumping different consensus flavors
+ * returns: 0 on success -1 on error. */
 STATIC int
-getinfo_helper_current_consensus(const char* question,
-                          char** answer,
-                          const char** errmsg)
+getinfo_helper_current_consensus(int flavor,
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: Instead of `int flavor` you can use the `consensus_flavor_t` enum.

File: src/feature/control/control_getinfo.c [hunk 33]
```diff
@@ -325,47 +325,37 @@ getinfo_helper_current_time(control_connection_t *control_conn,
...
+                                 const char** errmsg)
 {
-  // Ensures question contains a request for
-  // ns or microdesc consensus
-  if (
-      strcmp(question, "dir/status-vote/current/consensus") &&
-      strcmp(question, "dir/status-vote/current/consensus-microdesc")) {
-    return 0;
+  const char *flavor_name = networkstatus_get_flavor_name(flavor);
+  if (!strcmp(flavor_name, "")) {
```
Comment (asn-d6) [inline src/feature/control/control_getinfo.c]: This strcmp with "" is no longer useful after using `networkstatus_get_flavor_name`.  At least change it to check for "??" or change the general check logic.

File: src/test/test_controller.c [hunk 34]
```diff
@@ -1689,6 +1691,143 @@ test_download_status_bridge(void *arg)
...
+test_getinfo_helper_current_consensus_from_cache(void *arg)
+{
+  /* We just need one of these to pass, it doesn't matter what's in it */
+  control_connection_t dummy;
+  /* Get results out */
+  char *answer = NULL;
+  const char *errmsg = NULL;
+
+  (void)arg;
+  setup_bridge_mocks();
```
Comment (teor2345) [inline src/test/test_controller.c]: Not needed

File: src/test/test_controller.c [hunk 35]
```diff
@@ -1689,6 +1691,143 @@ test_download_status_bridge(void *arg)
...
+  getinfo_helper_dir(&dummy,
+                     "dir/status-vote/current/consensus-microdesc",
+                     &answer,
+                     &errmsg);
+  tt_str_op(answer, OP_EQ, "mock_microdesc_consensus");
+  tt_ptr_op(errmsg, OP_EQ, NULL);
+  tor_free(mock_microdesc_consensus_cache->dir);
+  tor_free(answer);
+  errmsg = NULL;
+  options->FetchUselessDescriptors = previous_fetch_value;
```
Comment (teor2345) [inline src/test/test_controller.c]: Should be after done:

File: src/test/test_controller.c [hunk 36]
```diff
@@ -1689,6 +1691,143 @@ test_download_status_bridge(void *arg)
...
+  /* We just need one of these to pass, it doesn't matter what's in it */
+  control_connection_t dummy;
+  /* Get results out */
+  char *answer = NULL;
+  const char *errmsg = NULL;
+
+  (void)arg;
+  setup_bridge_mocks();
+  or_options_t *options = get_options_mutable();
+  int previous_fetch_value = options->FetchUselessDescriptors;
```
Comment (teor2345) [inline src/test/test_controller.c]: Not needed

File: src/test/test_controller.c [hunk 37]
```diff
@@ -1689,6 +1691,143 @@ test_download_status_bridge(void *arg)
...
+                     &errmsg);
+  tt_str_op(answer, OP_EQ, "mock_microdesc_consensus");
+  tt_ptr_op(errmsg, OP_EQ, NULL);
+  tor_free(mock_microdesc_consensus_cache->dir);
+  tor_free(answer);
+  errmsg = NULL;
+  options->FetchUselessDescriptors = previous_fetch_value;
+
+ done:
+  clear_bridge_mocks();
```
Comment (teor2345) [inline src/test/test_controller.c]: Not needed

**Comment by teor2345** [comment]
Latest code in #1434.
